qiskit_dynamics/solvers/solver_utils.py

```diff
@@ -288,3 +292,102 @@ def setup_args_lists(
         x * max_len if arg_len == 1 else x for x, arg_len in zip(args_as_lists, arg_lens)
     ]
     return args_as_lists, args_were_lists
+
+
+class HashWrapper:
+    """Thin hashable wrapper over any object."""
+
+    def __init__(self, obj):
+        self.obj = obj
+        try:
+            self._hash = hash(obj)
+        except TypeError:
+            # force hash collisions on unhashable objects; whenever this HashWrapper
+            # is used in a hash table, it will have to fall back to __eq__
+            self._hash = 0
+
+    def __hash__(self):
+        return self._hash
+
+    def __eq__(self, other):
+        return self.obj == other
+
+    def __repr__(self):
+        return f"HashWrapper({repr(self.obj)})"
```

Comment:
This implementation assumes that once a `HashWrapper` instance is put into a hash map, the contents of `obj` is not mutated until the hash map leaves scope. For example, this is a valid use because the `set` is dereferenced before any objects can be mutated:

```python
def make_unique(objects):
    return list(set(map(HashWrapper, objects)))
```

However, this is problematic:

```python
def yield_unique(objects):
    yield from set(map(HashWrapper, objects))
```
because something like
```python
for obj in yield_unique(objects):
    obj.x = y
```
can cause undefined behaviour.

Therefore, I have a lingering concern over the whole premise of this PR because we are essentially in the second case: the hashmap of mutable objects lives inside of the `lru_cache` whose scope is ultimately out of our control.

Given these concerns, do we:
1) Accept the risk "because it probably won't matter"?
2) Modify `HashWrapper` to do potentially (expensive?) copies of obj?
3) Modify `HashWrapper` in some other complicated way?
4) Somehow tighten the scope of the hashmap?
5) Give up on the idea?

Reply:
As per our discussion earlier, these are the combinations of things that need to go into `HashWrapper`:
- `self`
    - I didn't think of this one initially when talking about this but I guess this is a bit annoying. I think we mostly think of the `Solver` as being immutable once constructed, and it doesn't explicitly have a public API for modifying it, but you _can_ change `solver.model`, e.g. by specifying a different rotating frame. It would be a shame to have to fully copy the solver. 
- `y_cls` 
    - This can be `None` (indicating its some form of array), `Statevector`, `DensityMatrix`, `SuperOp`, or `Operator` (where the last 4 are all from `qiskit.quantum_info`). Not sure if these types are hashable.
- `kwargs`
    - These are keyword arguments passed to `solve_lmde`, and this will only be triggered in the following cases.
    - If `kwargs['method'] == 'jax_odeint'`, then all of the other kwargs are going to be basic python types, like strings and numeric types. I think in this case there is no concern.
    - For the `diffrax` library, `kwargs['method']` will be an instance of a subclass of `AbstractSolver` (a `diffrax` class). An example usage is in [this test case](https://github.com/Qiskit/qiskit-dynamics/blob/85128b7fb7897be2450804dd2823215139167b02/test/dynamics/solvers/test_solver_functions.py#L617). Similarly, you can see in this example that the step size controlling arguments are set in a class called `PIDController`. There are other class-style arguments like this for diffrax solvers, like there is a `SaveAt` class for specifying time points at which the state should be saved in the results object.
    - The `diffrax` case complicates things, but I also think copying these should be fine, the ones I'm aware of are light-ish weight objects.